### PR TITLE
Reduce delays in controller tests

### LINKED ISSUE #1364: controller package tests take over 12s to complete
## Expected Behavior

Tests inside `./controller` take less than 5s to complete.

## Actual Behavior

They take about 12s to complete.

## Steps to Reproduce the Problem

```console
$ go test ./controller -count=1
ok      knative.dev/pkg/controller      12.590s
```

## Additional Info

I noticed this while working on #1363, will investigate later.

### PR DESCRIPTION
**Pull request by antoineco**
A series of tweaks to reduce unnecessary pauses in `controller` tests. Most of them related to excessive Sleeps.

Each commit corresponds to 1 test suite. See individual commit messages for details.

Closes #1364

### CODE REVIEW

**Comment by antoineco** [comment]
I'm moving this PR to "ready for review". We're down to ~5s versus >20s (both values from a run on my modest laptop) which sounds satisfying to me. There are still a few "slow" tests that were added after my original issue, so I'm gonna leave these untouched for now.

File: controller/controller_test.go [hunk 1]
```diff
@@ -771,59 +767,145 @@ func TestEnqueues(t *testing.T) {
...
-	// Sleep the remaining time.
-	time.Sleep(time.Second)
-	if got, want := impl.WorkQueue().Len(), 1; got != want {
-		t.Errorf("|Queue| = %d, want: %d", got, want)
+	}, shortDelay)
+
+	// Keep checking the queue length until 'to/fall' gets enqueued, send to channel to indicate success.
+	queuePopulated := make(chan struct{})
+	defer close(queuePopulated)
+	const queueCheckTimeout = shortDelay + 100*time.Millisecond
```
Comment (vagababov) [inline controller/controller_test.go:809]: put this up next to the other constants?

File: controller/controller_test.go [hunk 2]
```diff
@@ -771,59 +767,145 @@ func TestEnqueues(t *testing.T) {
...
+	ctx, cancel := context.WithTimeout(context.Background(), queueCheckTimeout)
+	defer cancel()
```
Comment (vagababov) [inline controller/controller_test.go:811]: `t.Cleanup`

File: controller/controller_test.go [hunk 3]
```diff
@@ -771,59 +767,145 @@ func TestEnqueues(t *testing.T) {
...
+					queuePopulated <- struct{}{}
+					return
+				}
+			}
+		}
+	}()
+
+	select {
+	case <-queuePopulated:
+		cancel()
```
Comment (vagababov) [inline controller/controller_test.go:828]: Do we need this cancel here? Presumably we left the gofunc already above?

File: controller/controller_test.go [hunk 4]
```diff
@@ -771,59 +767,145 @@ func TestEnqueues(t *testing.T) {
...
+		if enqueueDelay := time.Since(enqueueTime); enqueueDelay < shortDelay {
+			t.Errorf("Item enqueued within %s, expected at least a %s delay", enqueueDelay, shortDelay)
```
Comment (vagababov) [inline controller/controller_test.go:830]:
```suggestion
			t.Errorf("Item enqueued within %v, expected at least a %v delay", enqueueDelay, shortDelay)
```

File: controller/controller_test.go [hunk 5]
```diff
@@ -771,59 +767,145 @@ func TestEnqueues(t *testing.T) {
...
+		t.Fatal("Timed out waiting for item to be put onto the workqueue")
 	}
+
 	impl.WorkQueue().ShutDown()
-	if got, want := drainWorkQueue(impl.WorkQueue()), []types.NamespacedName{{Namespace: "the", Name: "waterfall"}}; !cmp.Equal(got, want) {
-		t.Errorf("Queue = %v, want: %v, diff: %s", got, want, cmp.Diff(got, want))
+
+	got, want := drainWorkQueue(impl.WorkQueue()), []types.NamespacedName{{Namespace: "to", Name: "fall"}}
+	if diff := cmp.Diff(want, got); diff != "" {
+		t.Error("Unexpected workqueue state (-:expect, +:got):\n", diff)
```
Comment (vagababov) [inline controller/controller_test.go:844]:
```suggestion
		t.Errorf("Unexpected workqueue state (-:expect, +:got):\n%s", diff)
```
Unfortunately all error calls insert a space, so for printing diffs we have to use formatters :(

File: controller/controller_test.go (hunk 3 above)
Comment (antoineco) [reply controller/controller_test.go:828]: You're right, if we don't cancel here the cleanup will take care of it. Removing it.

File: controller/controller_test.go (hunk 4 above)
Comment (antoineco) [reply controller/controller_test.go:830]: Done in 874e7058be0adef73f778eebb7a4c3768cd4de24

File: controller/controller_test.go (hunk 5 above)
Comment (antoineco) [reply controller/controller_test.go:844]: TIL, thanks for the tip.

File: controller/controller_test.go (hunk 3 above)
Comment (antoineco) [reply controller/controller_test.go:828]: Done in 18884e0548736e103758203e86ea0cdde43d02d2

File: controller/controller_test.go (hunk 2 above)
Comment (antoineco) [reply controller/controller_test.go:811]: Done in 7e3844a7c315ce4601dfac4be7054a56d647448e

File: controller/controller_test.go (hunk 1 above)
Comment (antoineco) [reply controller/controller_test.go:809]: Done in b419b391422e6ba32744e4284f67b570c6fa89ef

File: controller/controller_test.go (hunk 5 above)
Comment (antoineco) [reply controller/controller_test.go:846]: Done in 39b02449c4470b1ec1d41c011f452bf5957c1874

File: controller/controller_test.go [hunk 6]
```diff
@@ -771,59 +767,146 @@ func TestEnqueues(t *testing.T) {
...
-		t.Errorf("|Queue| = %d, want: %d", got, want)
+	}, shortDelay)
+
+	// Keep checking the queue length until 'to/fall' gets enqueued, send to channel to indicate success.
+	queuePopulated := make(chan struct{})
+	defer close(queuePopulated)
+	ctx, cancel := context.WithTimeout(context.Background(), queueCheckTimeout)
+	t.Cleanup(cancel)
+	go func() {
+		for {
```
Comment (vagababov) [inline controller/controller_test.go:816]: This is a very busy loop now -- basically 100% cpu worth loop :), do we want to add a sleep of a few ms in the default case?
(or rather replace `default:` with `case <-time.After(5*time.Millisecond)` ? Or may be even 3ms.

Comment (antoineco) [reply controller/controller_test.go:816]: [done](https://github.com/knative/pkg/compare/d7e119b0f8853d30430d7818994e003309ab1965..34bb8fdc86a02d470bfd5b57b3f94e2cdd3cd941)

**Comment by antoineco** [comment]
Rebased on `master`, let's see if it makes the `Build / Build (1.15.x, ubuntu-latest)` job happy.

**Comment by antoineco** [comment]
Would any of the controller approvers mind giving this PR some attention? @grantr @mattmoor @tcnghia

**Comment by mattmoor** [comment]
@vagababov probably has thoughts and has been in here most recently.

File: controller/controller_test.go [hunk 7]
```diff
@@ -771,59 +767,146 @@ func TestEnqueues(t *testing.T) {
...
+			case <-ctx.Done():
+				return
+			case <-time.After(5 * time.Millisecond):
+				if impl.WorkQueue().Len() > 0 {
+					queuePopulated <- struct{}{}
+					return
+				}
+			}
+		}
+	}()
```
Comment (mattmoor) [inline controller/controller_test.go:827]: There is a `wait.Until` that takes a context IIRC.

File: controller/controller_test.go [hunk 8]
```diff
@@ -771,59 +767,146 @@ func TestEnqueues(t *testing.T) {
...
```
Comment (mattmoor) [inline controller/controller_test.go:890]: same `wait.Until`

File: controller/controller_test.go [hunk 9]
```diff
@@ -1090,51 +1173,57 @@ func (er *ErrorReconciler) Reconcile(context.Context, string) error {
...
+				// can't fully reflect the real state of queue length.
+				// Here we need to wait for NumRequeues to be more than 1, to ensure
+				// the key get re-queued and reprocessed as expect.
+				if impl.WorkQueue().NumRequeues(item) > 1 {
+					itemRequeued <- struct{}{}
+					return
+				}
+			}
+		}
+	}()
```
Comment (mattmoor) [inline controller/controller_test.go:1215]: same `wait.Until`

File: controller/controller_test.go [hunk 10]
```diff
@@ -1090,51 +1173,57 @@ func (er *ErrorReconciler) Reconcile(context.Context, string) error {
...
+			select {
+			case <-ctx.Done():
+				return
+			case <-time.After(5 * time.Millisecond):
+				// Check that the work was requeued in RateLimiter, as NumRequeues
+				// can't fully reflect the real state of queue length.
+				// Here we need to wait for NumRequeues to be more than 1, to ensure
+				// the key get re-queued and reprocessed as expect.
+				if impl.WorkQueue().NumRequeues(item) > 1 {
+					itemRequeued <- struct{}{}
```
Comment (mattmoor) [inline controller/controller_test.go:1210]: If this is only expected to happen once, generally the pattern in Go is to `close()` the channel instead of sending the empty struct.  If it can happen multiple times, then this is cool too.

File: controller/controller_test.go (hunk 7 above)
Comment (antoineco) [reply controller/controller_test.go:827]: Sweet, thanks  for the tip!

File: controller/controller_test.go (hunk 10 above)
Comment (antoineco) [reply controller/controller_test.go:1210]: That's what I would usually do as well, but there is a dangerous `defer close(itemRequeued)` which causes a panic if I close the channel twice.

Do you think I should rather leave the channel open in case of failure and let it die with the process?

File: controller/controller_test.go (hunk 8 above)
Comment (antoineco) [reply controller/controller_test.go:890]: See previous comment.

**Comment by antoineco** [comment]
Reviewers, is anything blocking this PR in its current state?

File: controller/controller_test.go [hunk 11]
```diff
@@ -771,59 +768,138 @@ func TestEnqueues(t *testing.T) {
...
+			Namespace: "to",
 		},
-	}, 20*time.Second)
-	time.Sleep(10 * time.Millisecond)
-	if got, want := impl.WorkQueue().Len(), 0; got != want {
-		t.Errorf("|Queue| = %d, want: %d", got, want)
+	}, shortDelay)
+
+	// Keep checking the queue length until 'to/fall' gets enqueued, send to channel to indicate success.
+	queuePopulated := make(chan struct{})
```
Comment (vagababov) [inline controller/controller_test.go:812]: Do we want the channel to be buffered?

File: controller/controller_test.go [hunk 12]
```diff
@@ -771,59 +768,138 @@ func TestEnqueues(t *testing.T) {
...
+	defer close(queuePopulated)
```
Comment (vagababov) [inline controller/controller_test.go:813]: t.Cleanup as well? It's possible to join them with `func() {...}`.

File: controller/controller_test.go [hunk 13]
```diff
@@ -771,59 +768,138 @@ func TestEnqueues(t *testing.T) {
...
-		t.Errorf("Queue = %v, want: %v, diff: %s", got, want, cmp.Diff(got, want))
+
+	got, want := drainWorkQueue(impl.WorkQueue()), []types.NamespacedName{{Namespace: "to", Name: "fall"}}
+	if diff := cmp.Diff(want, got); diff != "" {
+		t.Errorf("Unexpected workqueue state (-:expect, +:got):\n%s", diff)
 	}
 }
 
-func TestEnqeueKeyAfter(t *testing.T) {
+func TestEnqueueKeyAfter(t *testing.T) {
```
Comment (vagababov) [inline controller/controller_test.go:847]: Same here.

File: controller/controller_test.go (hunk 11 above)
Comment (antoineco) [reply controller/controller_test.go:812]: What would be the benefit here? I rarely use buffered channels so bear with me.

File: controller/controller_test.go (hunk 12 above)
Comment (antoineco) [reply controller/controller_test.go:813]: Would the benefit of using it justify such an overloaded syntax? (again, bear with me, I've never used `t.Cleanup()` before this PR)

File: controller/controller_test.go [hunk 14]
```diff
@@ -771,59 +767,146 @@ func TestEnqueues(t *testing.T) {
 }
 
 func TestEnqueueAfter(t *testing.T) {
+	const (
+		// longDelay is longer than we expect the test to run.
+		longDelay = 1 * time.Minute
```
[the remark this answers is not in the document]
Comment (antoineco) [reply controller/controller_test.go]: Done in 105bdf7

File: controller/controller_test.go (hunk 12 above)
Comment (vagababov) [reply controller/controller_test.go:813]: If there's a problem, the test would fail cleanly after timeout, hopefully rather than timeout and panic because of the deadlock because channel consumer can't read

File: controller/controller_test.go (hunk 11 above)
Comment (vagababov) [reply controller/controller_test.go:812]: If there's a problem, the test would fail cleanly after timeout, hopefully rather than timeout and panic because of the deadlock because channel consumer can't read

Comment (antoineco) [reply controller/controller_test.go:812]: We never close the channel during the test execution (only after as a deferred action) so if I'm not mistaken there is no possible deadlock here. What am I missing?

File: controller/controller_test.go (hunk 12 above)
Comment (antoineco) [reply controller/controller_test.go:813]: Makes sense, on it.

File: controller/controller_test.go (hunk 11 above)
Comment (vagababov) [reply controller/controller_test.go:812]: Actually, I think I am overthinking this. It is not buffered part that will fail. But check below.

File: controller/controller_test.go [hunk 15]
```diff
@@ -771,59 +768,138 @@ func TestEnqueues(t *testing.T) {
...
+	case <-queuePopulated:
+		if enqueueDelay := time.Since(enqueueTime); enqueueDelay < shortDelay {
+			t.Errorf("Item enqueued within %v, expected at least a %v delay", enqueueDelay, shortDelay)
+		}
+		if got, want := impl.WorkQueue().Len(), 1; got != want {
+			t.Errorf("|Queue| = %d, want: %d", got, want)
+		}
+
+	case <-ctx.Done():
+		t.Fatal("Timed out waiting for item to be put onto the workqueue")
```
Comment (vagababov) [inline controller/controller_test.go:836]: Here is the problem, where if due to a bug the channel is not filled -- this will never exit. So we might want a timeout here as well.

Comment (antoineco) [reply controller/controller_test.go:841]: `ctx` is done after `shortDelay + 500*time.Millisecond` due to the `context.WithTimeout()` above, so `select` will find a candidate no matter what.
The test fails rather quickly here in case things go wrong, I've tried to do things like short circuiting the channel writer while refactoring those tests, and I'm pretty confident the edge cases mentioned are all covered here.

Comment (vagababov) [reply controller/controller_test.go:841]: 😊 
Ah, nice.
Thanks for pointing out. We rarely do with timeout so it elided me.

Comment (antoineco) [reply controller/controller_test.go:841]: Yeah those tests are very much time-based due to the introduction of arbitrary delays ("EnqueueAfter"). I would love to use something more elegant than a timeout, it's just to prevent those tests from running forever.

**Comment by antoineco** [comment]
Appreciate the final thorough review @vagababov 👍
